[2.19] Concurrency updates for isolate.run() (#4357)
dart-lang/site-www: src/_guides/language/concurrency/index.md

[`send()` method]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-isolate/SendPort/send.html

Message passing normally involves data copying, and thus can be slow.
However, when you send data using isolates,
the memory that holds the message in the worker isolate isn’t copied,

> but instead is transferred to the receiving isolate.

Reviewer: This is not generally true (and not sure what "send data using isolates" mean). It's only true for the message sent using `Isolate.exit`.
So, should it say: "When you send data using `Isolate.exit`,"?

Author: That's correct, this paragraph was previously under the first example and specifically about `SendPort.send` vs. `Isolate.exit()`. I was trying to generalize it and make it a conceptual summary instead of the specific mention of those methods because they're abstracted now. But, I don't want to completely remove the info...

I'll see how I can incorporate the valuable parts of this more accurately and then request your review again, thanks!

The sender will nonetheless perform a verification pass to ensure

> the objects are allowed to be transferred.

Reviewer: (That does seem to match `Isolate.exit` behavior.)

 [Flutter `compute()` function]: {{site.flutter-docs}}/cookbook/networking/background-parsing#4-move-this-work-to-a-separate-isolate

### Implementing a simple worker isolate

> These examples show the implementation of a
> main isolate that spawns a simple worker isolate.
> The examples use [`Isolate.run()`][], 
> which efficiently abstracts the complexity
> of setting up and managing worker isolates manually:

Reviewer 2: **suggestion:** Remove the nominalizations and adverbs. Both make the copy wordy and distract some from the understanding.

```suggestion
These examples implement a main isolate that spawns
a simple worker isolate.
Using [`Isolate.run()`][] simplifies setting up and
managing worker isolates.
```

> * Spawns (starts and creates) an isolate

Reviewer 2: **question:** Are these items in order? If so, should they not be an ordered list?

Author: Yes, they are ordered. Will do!

> * Takes a function argument and executes

Reviewer 3: This was hard for me to parse. These are steps that Isolate.run does for you right? It also sounds like we're saying it takes a function argument _after_ the isolate is spawned. 

Instead I'd probably say something like:
```suggestion
* Runs a function on the spawned isolate
```

* Captures the result
* Returns the result to the main isolate
* Terminates the isolate once work is complete
* Checks, captures, and throws exceptions and errors back to the main isolate

[`Isolate.run()`]: {{site.dart-api}}/dev/dart-isolate/Isolate/run.html

#### Running an existing method in a new isolate

> Here’s the code for the main isolate:

Reviewer 2: **suggestion:** Don't use "here" to indicate location. "Here" would be that exact line. It's not "here", the code _follows_ this block. I know that it's good to have copy before an example and after a heading, but make sure it adds value.
```suggestion
You can write a main isolate that resembles the following code:
```

**question:** What makes this the "main" isolate? Is it because it's in the `main` class? Does `main` = "first"? I would recommend clarifying this here and under **The main isolate**.

Author: The main isolate is the isolate that every Dart program run in by default, so yes, the "first" or "starting" (defined higher up [on the same page](https://dart.dev/guides/language/concurrency#the-main-isolate)). I'll go through the page and make sure the terminology is consistent, thanks for putting it on my radar!

<?code-excerpt "lib/simple_worker_isolate.dart (main)"?>
```dart
void main() async {
  // Read some data.
  final jsonData = await Isolate.run(_readAndParseJson);

  // Use that data.
  print('Number of JSON keys: ${jsonData.length}');
}
```

> The spawned isolate executes the following code:

Reviewer 2: **question:** Is this a spawned isolate or the main isolate? If it's the main, you may want to update this.

Author: This is the spawned one, spawned by the main isolate.

> The spawned isolate executes the following code:

Reviewer 3: nitpick:
```suggestion
The spawned isolate executes the function passed as the first argument, `_readAndParseJson`:
```

<?code-excerpt "lib/simple_worker_isolate.dart (spawned)"?>
```dart
Future<Map<String, dynamic>> _readAndParseJson() async {
  final fileData = await File(filename).readAsString();
  final jsonData = jsonDecode(fileData) as Map<String, dynamic>;
  return jsonData;
}
```

1. `Isolate.run()` spawns an isolate for the background worker,
   while `main()` waits for the result.

2. The new isolate executes the argument passed to `run()`:

>    the function `_readAndParseJson()`.

Reviewer 2: **suggestion:** Recommend providing some consistency: "...spawns a new isolate..." and "The new isolate..." _or_ "...spawns an isolate..." and "The spawned isolate...".

Reviewer 2: **question:** Reading the next step, when is the worker isolate created? Is that different than the main? It's unclear here.

> 3. The `return` statement ends the computation,
>    shutting down the worker isolate and
>    sending `jsonData` back to the main isolate as the result.

Reviewer 2: **suggestion:** Tighten the language somewhat.
```suggestion
3. The `return` statement shuts down the spawned isolate and
   sends `jsonData` back to the main isolate as the result.
```

Reviewer 3: It's a little pedantic, but `return` doesn't shut down the isolate, Isolate.run() will take this result and call [exit](https://api.dart.dev/dev/2.19.0-388.0.dev/dart-isolate/Isolate/exit.html), which sends the value back to the main isolate.

> 4. The memory holding the result in the worker isolate is *transferred*
>    to the main isolate, *not copied*.
>    The worker isolate still performs a verification pass to ensure
>    the objects are allowed to be transferred.

Reviewer 2: **suggestion:** Clarify who has agency. Is it the worker isolate? If so, use the active voice here. Also, why 

Also, clarify what the work isolate is _doing_.
```suggestion
4. The worker isolate transfers the heap holding the result
   to the main isolate. It _does not copy_ the heap.
   The worker isolate verifies if it can transfer the objects.
```

Author: Incorporated these suggestions, except:
> Also, why

I think it's reasonable to assume the concept of _transferring_ memory rather than _copying_ data is an understood benefit in general software engineering (maybe I'm wrong?). I don't want to clutter the page by explaining it. Also, the rest of the examples don't get into the motivation behind the implementation details of isolates, so I'd rather avoid that here. That kind of background is more blog post material.

> `_readAndParseJson()` is an existing,
> asynchronous function that could just as easily
> run directly in the main isolate.
> But using `Isolate.run()` enables concurrency,
> so its computations are completely abstracted by the worker isolate
> and can complete without blocking the main isolate.

Reviewer 2: **suggestion:** Tighten the language. Use active voice.
```suggestion
The existing `_readAndParseJson()` asynchronous function
could run within the main isolate.
Use `Isolate.run()` to enable concurrency.
This allows the worker isolate to abstract computations
and complete the work without blocking the main isolate.
```

Author: I kept the first sentence as-is because the emphasis of the example is that you're just running normal functions you could use anywhere. I feel like the suggestion loses a bit of that emphasis. Tightened the rest though!

> Because `Isolate.run()` allows code in the main isolate to continue running
> while its computations execute in a new one,
> its result is always returned as a `Future`.
> The original isolate doesn't care whether the computation
> running in the other isolate is synchronous or asynchronous,
> because it's being run concurrently either way.

Reviewer 2: **suggestion:** This is a matter of tightening language again. Shorten the sentences, use active voice, avoid anthropomorphizing.
```suggestion
The `Isolate.run()` method returns a `Future`. 
This result returns because the main isolate continues to run
while computations execute in the new isolate.
The code running in the new isolate doesn't impact
 the original isolate.
```

{% comment %}
TODO:
Should create a diagram for the current example.
Previous example's diagram and text for reference:

  The following figure illustrates the communication between
  the main isolate and the worker isolate:
  
  ![A figure showing the previous snippets of code running in the main isolate and in the worker isolate](/guides/language/concurrency/images/isolate-api.png)
{% endcomment %}

#### Sending closures with isolates

> This section shows another simple worker isolate created with `run()`,
> this time written as a function literal, or closure,
> directly in the main isolate.

Reviewer 2: **suggestion:** Avoid the "This section..." trap.
```suggestion
You can also write a simple worker isolate a
function literal, or closure, in the main isolate.
```

<?code-excerpt "lib/simple_isolate_closure.dart (main)"?>
```dart
void main() async {
  // Read some data.
  final jsonData = await Isolate.run(() async {
    final fileData = await File(filename).readAsString();
    final jsonData = jsonDecode(fileData) as Map<String, dynamic>;
    return jsonData;
  });

  // Use that data.
  print('Number of JSON keys: ${jsonData.length}');
}
```

> This example accomplishes the same as the previous:
> a new isolate spawns and executes a computation, sending back the result.

Reviewer 2: **suggestion:** Avoid a compound sentence with a colon. If you can use a colon, you can split the sentence.
```suggestion
This example accomplishes the same as the previous.
A new isolate spawns, executes code, and returns the result.
```

> However, now the isolate sends a [closure][]. 
> This allows versatility not only for how the isolate can be utilized
> (as closures are less limited than typical named functions),
> but also for the written style of the code.
> Here, `Isolate.run()` executes what looks like local code, concurrently.
> In that sense, you can imagine `run()` almost like a [control flow operator][]
> for “run in parallel”.

Reviewer 2: **suggestion:** This could also use tightening.

```suggestion
Now the isolate returns a [closure][]. 
Closures provide more options as to how to use the isolate
and how to write the code.
They have fewer limits than named functions.
In this example, `Isolate.run()` executes what looks like
local code in a concurrent manner.
In that sense, imagine `run()` to work like a [control flow operator][]
for “run in parallel”.
```

Author: Is it a hard-must to get rid of every "However"? I do overuse it sometimes, but I feel like in this case it's somewhat warranted... 

Reading the line preceding this paragraph, and then following with "Now" feels abrupt and reads like an incomplete thought. 

> This example accomplishes the same as the previous. A new isolate spawns, computes something, and sends back the result.
> 
> Now the isolate sends a closure

Like, it doesn't sound like we're talking about the difference between the previous example and this one.
I want to keep the emphasis on the _difference_, not just stating what's happening for the sake of stating it. Is there another way to do that "feels" smoother?

[closure]: /guides/language/language-tour#anonymous-functions
[control flow operator]: /guides/language/language-tour#control-flow-statements

### Sending multiple messages between isolates

`Isolate.run()` abstracts a handful of lower-level, 
isolate-related API to simplify isolate management:

* [`Isolate.spawn()`][] and [`Isolate.exit()`][]
* [`ReceivePort`][] and [`SendPort`][]

[`Isolate.exit()`]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-isolate/Isolate/exit.html
[`Isolate.spawn()`]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-isolate/Isolate/spawn.html
[`ReceivePort`]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-isolate/ReceivePort-class.html
[`SendPort`]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-isolate/SendPort-class.html

> However, there are cases where it’s necessary
> to directly utilize the primitives that make up `run()`
> for more granular control over isolate management.
> Sending multiple messages between isolates is one of those cases.

Reviewer 2: **suggestion:** Don't bury the lede. Also, simplify the language. (That is, "control" and "management" are synonyms.)
```suggestion
To provide more granular control of isolates might
require using the primitives that make up `run()`.
One example use case would be sending multiple
messages between isolates.
```

For more communication between isolates,
you need to use the [`send()` method][] of `SendPort`.
One common pattern, which the following figure shows,
is for the main isolate to send a request message to the worker isolate,

> which then sends one or more reply messages.

Reviewer 2: **question:** What does "for _more_ communication" mean? I would recommend finding a way to proffer an example as part of the explanation.

Author: The section header is "Sending multiple messages between isolates", that's what "more communication" is in reference to. The context is that one message is the standard, more than one requires different configuration. I'll figure out how to include that

### Implementing a simple worker isolate

These examples show the implementation of a
main isolate that spawns a simple worker isolate.
The examples use [`Isolate.run()`][], 
which efficiently abstracts the complexity
of setting up and managing worker isolates manually:

* Spawns (starts and creates) an isolate
* Takes a function argument and executes
* Captures the result
* Returns the result to the main isolate
* Terminates the isolate once work is complete
* Checks, captures, and throws exceptions and errors back to the main isolate

[`Isolate.run()`]: {{site.dart-api}}/dev/dart-isolate/Isolate/run.html

#### Running an existing method in a new isolate

Here’s the code for the main isolate:

<?code-excerpt "lib/simple_worker_isolate.dart (main)"?>
```dart
void main() async {
  // Read some data.
  final jsonData = await Isolate.run(_readAndParseJson);

>   // Use that data

Reviewer: (Final `.`).

  print('Number of JSON keys: ${jsonData.length}');
}
```

The spawned isolate executes the following code:

<?code-excerpt "lib/simple_worker_isolate.dart (spawned)"?>
```dart
Future<Map<String, dynamic>> _readAndParseJson() async {
  final fileData = await File(filename).readAsString();
  final jsonData = jsonDecode(fileData) as Map<String, dynamic>;
  return jsonData;
}
```

1. `Isolate.run()` spawns an isolate for the background worker,
   while `main()` waits for the result.

2. The new isolate executes the argument passed to `run()`:
   the function `_readAndParseJson()`.

3. The `return` statement ends the computation,
   shutting down the worker isolate and
   sending `jsonData` back to the main isolate as the result.

In this example, `_readAndParseJson()` is an existing,
asynchronous function that could just as easily
run directly in the main isolate.
But using `Isolate.run()` enables concurrency,
so its computations are completely abstracted by the worker isolate
and can complete without blocking the main isolate.

> An added benefit is that the result of

Reviewer: Being asynchronous is not what people would normally call a "benefit" :)

I'd say rather just not value-load it, and say something like:

> Because `Isolate.run()` allows other code to run while the computation
> executed in another isolate, its result is always returned as a `Future`.
> To the original isolate, it doesn't matter whether the computation running in the 
> other isolate is synchronous or asynchronous, because it's being run concurrently
> either way.

* Maybe:
  * Add a new macro & style for flutter notes?
  * Add the following text somewhere in this page (or in the FAQ):

>     * Sometimes Dart is called a _single-threaded language._

Reviewer 3: Are the rest of these TODOs still relevant?

Author: No, not for me at least. I'll incorporate the "Dart code executes..." line then remove the whole comment in my next edit.

Reviewer 4: Removing the TODOs and maybes seems good to me! The important ones are fixed or tracked already.
